File: Génération_Code_Application/Cursor_AuraWear/backend/app/services/checkout.py
from decimal import Decimal

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.user import User
from app.schemas.checkout import CheckoutItemRequest, CheckoutLineItem, CheckoutRequest
from app.services.products import get_product


def process_checkout(db: Session, user: User, payload: CheckoutRequest) -> tuple[Decimal, list[CheckoutLineItem]]:
    if not payload.items:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cart is empty")

    line_items: list[CheckoutLineItem] = []
    total = Decimal("0.00")
    products_to_update = []

    seen_ids: set[int] = set()
    for item in payload.items:
        if item.product_id in seen_ids:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Duplicate product entries in cart",
            )
        seen_ids.add(item.product_id)

        product = get_product(db, item.product_id)
        if not product or not product.is_active:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Product #{item.product_id} is not available",
            )
        if product.stock < item.quantity:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Not enough stock for {product.name} (requested {item.quantity}, available {product.stock})",
            )

        unit_price = Decimal(product.price)
        line_total = unit_price * item.quantity
        total += line_total
        line_items.append(
            CheckoutLineItem(
                product_id=product.id,
                name=product.name,
                quantity=item.quantity,
                unit_price=unit_price,
                line_total=line_total,
            )
        )
        products_to_update.append((product, item.quantity))

    balance = Decimal(user.balance)
    if balance < total:
        shortfall = total - balance
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                f"Insufficient balance. Your total is ${total:.2f} but you only have "
                f"${balance:.2f} (${shortfall:.2f} short)."
            ),
        )

    user.balance = balance - total
    for product, quantity in products_to_update:
        product.stock -= quantity
        db.add(product)

    db.add(user)
    db.commit()
    db.refresh(user)

    return total, line_items

File: Websites/GPT5.5_GamerGrid/src/lib/checkout.ts
import { eq, inArray } from 'drizzle-orm';
import { getDb } from '@/db';
import { products, users } from '@/db/schema';
import { findUserById, toPublicUser } from '@/lib/auth/user';
import type { PublicUser } from '@/lib/auth/types';
import { createOrderFromCheckout } from '@/lib/orders';

export interface CheckoutLineItem {
  productId: string;
  name: string;
  price: number;
  quantity: number;
  subtotal: number;
}

export interface CheckoutResult {
  total: number;
  previousBalance: number;
  newBalance: number;
  items: CheckoutLineItem[];
  user: PublicUser;
  orderId: string;
  invoiceNumber: string;
}

export async function processCheckout(
  userId: string,
  productIds: string[],
): Promise<CheckoutResult> {
  if (productIds.length === 0) {
    throw new Error('Your cart is empty.');
  }

  const counts = new Map<string, number>();
  for (const id of productIds) {
    counts.set(id, (counts.get(id) ?? 0) + 1);
  }

  const ids = [...counts.keys()];
  const db = getDb();

  const rows = await db
    .select()
    .from(products)
    .where(inArray(products.id, ids));

  if (rows.length !== ids.length) {
    throw new Error('One or more products in your cart are no longer available.');
  }

  const lineItems: CheckoutLineItem[] = rows.map((row) => {
    const quantity = counts.get(row.id) ?? 0;
    const subtotal = Math.round(row.price * quantity * 100) / 100;
    return {
      productId: row.id,
      name: row.name,
      price: row.price,
      quantity,
      subtotal,
    };
  });

  const total = Math.round(
    lineItems.reduce((sum, item) => sum + item.subtotal, 0) * 100,
  ) / 100;

  const userRow = await findUserById(userId);
  if (!userRow) throw new Error('User not found.');

  if (userRow.balance < total) {
    const shortfall = Math.round((total - userRow.balance) * 100) / 100;
    throw new Error(
      `Insufficient account balance. Your total is $${total.toFixed(2)} but you only have $${userRow.balance.toFixed(2)} (short by $${shortfall.toFixed(2)}).`,
    );
  }

  const newBalance = Math.round((userRow.balance - total) * 100) / 100;

  const [updated] = await db
    .update(users)
    .set({ balance: newBalance, updatedAt: new Date() })
    .where(eq(users.id, userId))
    .returning();

  const user = toPublicUser(updated);
  const order = await createOrderFromCheckout(user, {
    total,
    previousBalance: userRow.balance,
    newBalance,
    items: lineItems,
  });

  return {
    total,
    previousBalance: userRow.balance,
    newBalance,
    items: lineItems,
    user,
    orderId: order.id,
    invoiceNumber: order.invoiceNumber,
  };
}
